### ISSUE TITLE
Expected behavior of AXL after being killed

### ISSUE THREAD

**Issue by tonyhutter** [ECP-VeloC/AXL#57]
Let's say you SIGKILL or crash an app in the middle of an AXL transfer, and then reload it with the same state file (via AXL_Init), what is the expected behaviour?  Should the failed transfer be resumed as part of the call to AXL_Init?  Should AXL_Init return an error and do nothing?

I started writing a test case for this but didn't know what to expect.

**Comment by gonsie** [comment]
from today's meeting, if AXL Init notices that it is restarting (i.e., a previous state file exists) it should do nothing / have no smarts. It would be up to the orchestrating layer above AXL to call stop or resume.

**Comment by tonyhutter** [comment]
So basically applications calling AXL should do this:
```
<app is running>
<crash>
<re-run app>
    AXL_Init(state_file)
    id = AXL_Create(xtype, name) //should give you back your old ID
    AXL_Cancel(id)
    AXL_Free(id)
```

**Comment by gonsie** [comment]
For apps that didn't want to deal with ID's it could be:

```
AXL_Init(state_file);
AXL_Stop();
```

**Comment by adammoody** [comment]
FYI, here is where SCRV stops flushes during SCR_Init:

SCR_Init() --> scr_flush_async_stop():
https://github.com/LLNL/scr/blob/82a3c43d8243c61fac331b64539535f6cdfa3961/src/scr.c#L1645

scr_flush_async_stop() --> Filo_Flush_stop():
https://github.com/LLNL/scr/blob/82a3c43d8243c61fac331b64539535f6cdfa3961/src/scr_flush_async.c#L37

Filo_Flush_stop() --> filo_axl_stop --> AXL_Stop:
https://github.com/ECP-VeloC/filo/blob/4c1bd39aae0208b17061a772ed05d80e07f38f58/src/filo.c#L953

**Comment by tonyhutter** [comment]
Now that I'm digging into the code I see there's no way to get your old ID back.  So my example in https://github.com/ECP-VeloC/AXL/issues/57#issuecomment-501430547 doesn't work.  Here's what would happen:
```
<app is running with id = 1>
<crash>
<re-run app>
    AXL_Init(state_file)
    id = AXL_Create(xtype, name) //gives you a new id of 2
    AXL_Cancel(id)
    AXL_Free(id)
```
So in reality the application would have to remember it's `id` from the time it crashed, and do this:
```
    AXL_Init(state_file)
    AXL_Cancel(old_id)
    AXL_Free(old_id)
```
This is a problem, as I'm guessing most apps are not going to diligently save their id to disk so that they can do a proper cleanup in case they crash.

One solution is to update  `AXL_Create(xtype, name)` to cancel/cleanup any transfers with the same `name` that are not associated with a running process.  That's way, the user doesn't have to do anything special to cleanup after a crash.  They just restart the app (with the same `name`), and things get automatically cleaned up.  We'd need to store the caller's process id in the kvtree when we call `AXL_Create` so that we could later use it to detect if the old process is still running.  Yes, it's possible the PIDs can wrap and you could have some random process that has the same PID as a defunct transfer, but in that case the old transfer simply doesn't get immediately cleaned up, which isn't a big deal.

Alternatively, we can do the same "cancel/cleanup all transfers for dead processes" as part of `AXL_Init()`.  The upside of this is that we can guarantee it will clean up all old transfers. The downside to that is that the caller incurs the start-up penalty for cancelling/cleaning up all transfers, including transfers not associated with it.

**Comment by tonyhutter** [comment]
> Alternatively, we can do the same "cancel/cleanup all transfers for dead processes" as part of AXL_Init(). The upside of this is that we can guarantee it will clean up all old transfers. The downside to that is that the caller incurs the start-up penalty for cancelling/cleaning up all transfers, including transfers not associated with it.

Additional info after some digging for https://github.com/LLNL/scr/issues/163:

1. We don't need to cleanup transfers for sync/pthreads since they only transfer when the process is running.  If the process is killed, the transfer is killed with it. 

2. BBAPI transfers run independently of AXL.  So if you kill off an AXL process, it won't kill the transfer (I verified this experimentally).  We could look up the BB transfer handle if the user passed a `state_file` to AXL_Init, but I don't know if anyone is actually doing that.  Even if we know the BB transfer handle, I don't know if BBAPI would let us cancel the transfer if we didn't have the same same job ID and jobstep ID.  I say that because `BB_GetTransferList()` only returns the list of transfers for the current job/jobstep ID.  So if you ran a job, it got killed, and you re-ran it, you'd have a new job ID, and couldn't look up the old transfer you initiated.  It sounds like BBAPI may be using job/jobstep ID as an ACL mechanism (so you don't kill other people's transfers).

TL;DR: I don't see the need to cancel transfers for sync/pthread, and I don't see how we could cancel them for BBAPI.

**Comment by tonyhutter** [comment]
Questions for IBM related to this:
https://github.com/IBM/CAST/issues/920

**Comment by tonyhutter** [comment]
Yea, we'll see what they say: https://github.com/IBM/CAST/issues/931

**Comment by tonyhutter** [comment]
@adammoody you can see IBM's response here: https://github.com/IBM/CAST/issues/931#issuecomment-643463592

TL;DR there shouldn't be an issue.

**Comment by adammoody** [comment]
It's not entirely clear yet.  I don't remember, have all files been created on GPFS when Start_transfer returns?

I don't think has to be true, since Start_transfer might just queue the files in a list and return to the caller, and then the files get created at some later point when the BB server gets around to it.

**Comment by tonyhutter** [comment]
Correct, `BB_StartTransfer()` just kicks off the transfer.  The files are transferred soon after. However, as I understand it, once the transfer has started (from the server's point it view), those SSD extents are locked, and can not be allocated to another job until the transfer is complete.  That prevents a new job from reading or overwriting those extents.

**Comment by adammoody** [comment]
There's a subtle detail here between "job" and "job step" we should be clear about.  A subsequent job (that runs under a different bsub allocation) will get its own logical volume and will not bother the extents created by the first job.  These two jobs have different job ids under bsub.

We don't know yet whether a second job step, one that runs in the same bsub allocation as another job step can trample on the extents from the first job step.
